Assets sometimes fail to load from the asset cache.

**Describe the bug**
When adding many new tokens using assets from the resource library, it sometimes happens that the asset is not fully read and results in MapTools freezing.

**To Reproduce**
Unfortunately I was not able to reproduce this without 5e Framework since I'm not sure exactly what it's doing to load the assets.

Steps to reproduce the behavior:
1. Clean MapTool installation
2. Downloads Rod's 5e Framework
3. Install the image files from https://get.5e.tools/.
4. Open up the Creature List and clicking "Make Token" for several different creatures.
5. After MapTool freezes, force quit.
6. Load up the campaign again and try to add the same creature that resulted in a freeze. This time it will work.

**Expected behavior**
MapTool should load the asset without freezing.

**MapTool Info**
- Version: 1.8.3, reproduced on latest develop (7664101b1a426c68f76723bf57a8db61e5f4eccb).
- Install: New

**Desktop (please complete the following information):**
 - OS: Linux Mint
 - Version 20

**Additional context**
I'm not sure if MapTool itself is freezing, or if the framework is somehow holding it up. I also don't know how to get the logs from a frozen MapTool process, so I reproduced the issue with the latest on `develop` and got these logs right when the freeze occurred:
```
14:49:11.392 (AssetManager.java:454) [pool-4-thread-1] ERROR net.rptools.maptool.model.AssetManager - MD5 for asset Camel corrupted
14:49:11.397 (ImageManager.java:428) [pool-8-thread-1] ERROR net.rptools.maptool.util.ImageManager - BackgroundImageLoader.run(Camel,png, d453500b4f82966f1d166b65acaa1c12): image not resolved
javax.imageio.IIOException: Error reading PNG image data
	at com.sun.imageio.plugins.png.PNGImageReader.readImage(PNGImageReader.java:1523) ~[?:?]
	at com.sun.imageio.plugins.png.PNGImageReader.read(PNGImageReader.java:1840) ~[?:?]
	at javax.imageio.ImageIO.read(ImageIO.java:1468) ~[?:?]
	at javax.imageio.ImageIO.read(ImageIO.java:1363) ~[?:?]
	at net.rptools.lib.image.ImageUtil.bytesToImage(ImageUtil.java:302) ~[main/:?]
	at net.rptools.maptool.util.ImageManager$BackgroundImageLoader.run(ImageManager.java:424) [main/:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
	at java.lang.Thread.run(Thread.java:831) [?:?]
Caused by: java.io.EOFException: Unexpected end of ZLIB input stream
	at java.util.zip.InflaterInputStream.fill(InflaterInputStream.java:244) ~[?:?]
	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158) ~[?:?]
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:244) ~[?:?]
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:284) ~[?:?]
	at java.io.BufferedInputStream.read(BufferedInputStream.java:343) ~[?:?]
	at java.io.DataInputStream.readFully(DataInputStream.java:201) ~[?:?]
	at com.sun.imageio.plugins.png.PNGImageReader.decodePass(PNGImageReader.java:1245) ~[?:?]
	at com.sun.imageio.plugins.png.PNGImageReader.decodeImage(PNGImageReader.java:1381) ~[?:?]
	at com.sun.imageio.plugins.png.PNGImageReader.readImage(PNGImageReader.java:1514) ~[?:?]
	... 8 more
```
I did some digging and added some additional logging to AssetManager, and what I found is that [this call to `FileUtils.readFileToByteArray()`](https://github.com/RPTools/maptool/blob/develop/src/main/java/net/rptools/maptool/model/AssetManager.java#L448) does not return all the data for the asset. This makes the MD5 to be incorrect and causes the inflation to fail.

--- a/src/main/java/net/rptools/maptool/model/AssetManager.java
+++ b/src/main/java/net/rptools/maptool/model/AssetManager.java
@@ -23,6 +23,9 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.net.URL;
+import java.nio.file.AtomicMoveNotSupportedException;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -82,6 +85,7 @@ public class AssetManager {
   private static AssetLoader assetLoader = new AssetLoader();
 
   private static ExecutorService assetLoaderThreadPool = Executors.newFixedThreadPool(1);
+  private static ExecutorService assetWriterThreadPool = Executors.newFixedThreadPool(1);
 
   static {
     cacheDir = AppUtil.getAppHome("assetcache");
@@ -451,7 +455,9 @@ public class AssetManager {
       Asset asset = new Asset(props.getProperty(NAME), data);
 
       if (!asset.getId().equals(id)) {
-        log.error("MD5 for asset " + asset.getName() + " corrupted");
+        log.error("MD5 for asset " + asset.getName() + " corrupted; purging corrupted file");
+        assetFile.delete();
+        return null;
       }
 
       assetMap.put(id, asset);
@@ -525,23 +531,24 @@ public class AssetManager {
     }
 
     if (!assetIsInPersistentCache(asset)) {
-
       final File assetFile = getAssetCacheFile(asset);
 
-      new Thread(
-              () -> {
-                assetFile.getParentFile().mkdirs();
-                try (OutputStream out = new FileOutputStream(assetFile)) {
-                  out.write(asset.getImage());
-                } catch (IOException ioe) {
-                  log.error("Could not persist asset while writing image data", ioe);
-                  return;
-                } catch (NullPointerException npe) {
-                  // Not an issue, will update once th frame is finished loading...
-                  log.warn("Could not update statusbar while MapTool frame is loading.", npe);
-                }
-              })
-          .start();
+      assetWriterThreadPool.submit(
+          () -> {
+            assetFile.getParentFile().mkdirs();
+
+            try (var operation = new AssetWriteRenameOperation(assetFile);
+                var temporaryFileStream = new FileOutputStream(operation.temporaryFile)) {
+              temporaryFileStream.write(asset.getImage());
+              // Now that the data is in a file, we move it to its final resting place.
+              operation.commit();
+            } catch (IOException ioe) {
+              log.error("Could not persist asset while writing image data", ioe);
+            } catch (NullPointerException npe) {
+              // Not an issue, will update once th frame is finished loading...
+              log.warn("Could not update statusbar while MapTool frame is loading.", npe);
+            }
+          });
     }
     if (!assetInfoIsInPersistentCache(asset)) {
 
@@ -815,5 +822,48 @@ public class AssetManager {
       missing.put(entry.getKey(), entry.getValue());
     }
     return missing;
+  }
+
+  /** Helper type to handle creating and moving temporary files. */
+  private static class AssetWriteRenameOperation implements AutoCloseable {
+    private final File assetFile;
+    private final File temporaryFile;
+
+    public AssetWriteRenameOperation(File assetFile) throws IOException {
+      this.assetFile = assetFile;
+      // Placing the temp file in the cache dir means it will be on the same filesystem in typical
+      // cases.
+      this.temporaryFile =
+          Files.createTempFile(assetFile.getParentFile().toPath(), "tmp.", "").toFile();
+    }
+
+    /**
+     * Move the temporary file to its final location.
+     *
+     * <p>The move will be done atomically if possible, but a non-atomic move may be used as a
+     * fallback.
+     *
+     * @throws IOException If the move fails.
+     */
+    public void commit() throws IOException {
+      try {
+        Files.move(temporaryFile.toPath(), assetFile.toPath(), StandardCopyOption.ATOMIC_MOVE);
+      } catch (AtomicMoveNotSupportedException e) {
+        Files.move(temporaryFile.toPath(), assetFile.toPath());
+      }
+    }
+
+    /**
+     * Clean up the temporary file.
+     *
+     * <p>If the operation has been committed, there is no longer a temporary file and this does
+     * nothing.
+     *
+     * @throws IOException
+     */
+    @Override
+    public void close() throws IOException {
+      Files.deleteIfExists(temporaryFile.toPath());
+    }
   }
 }

